SimPEG/PF/Gravity.py

```diff
@@ -19,120 +19,38 @@ class GravityIntegral(Problem.LinearProblem):
-
-                if self.actInd.dtype == 'bool':
-                    inds = np.asarray([inds for inds,
-                                      elem in enumerate(self.actInd, 1)
-                                      if elem], dtype=int) - 1
-                else:
-                    inds = self.actInd
-
-            else:
-
-                inds = np.asarray(range(self.mesh.nC))
-
-            nC = len(inds)
-
-            # Create active cell projector
-            P = sp.csr_matrix(
-                (np.ones(nC), (inds, range(nC))),
-                shape=(self.mesh.nC, nC)
-            )
-
-            # Create vectors of nodal location
-            # (lower and upper corners for each cell)
-            xn = self.mesh.vectorNx
-            yn = self.mesh.vectorNy
-            zn = self.mesh.vectorNz
-
-            yn2, xn2, zn2 = np.meshgrid(yn[1:], xn[1:], zn[1:])
-            yn1, xn1, zn1 = np.meshgrid(yn[0:-1], xn[0:-1], zn[0:-1])
-
-            Yn = P.T*np.c_[Utils.mkvc(yn1), Utils.mkvc(yn2)]
-            Xn = P.T*np.c_[Utils.mkvc(xn1), Utils.mkvc(xn2)]
-            Zn = P.T*np.c_[Utils.mkvc(zn1), Utils.mkvc(zn2)]
-
-            rxLoc = self.survey.srcField.rxList[0].locs
-            ndata = rxLoc.shape[0]
-
-            # Pre-allocate space and create magnetization matrix if required
-            # Pre-allocate space
-            if self.rtype == 'z':
-
-                fwr_d = np.zeros(self.survey.nRx)
-
-            elif self.rtype == 'xyz':
+            # Compute the linear operation without forming the full dense G
+            fields = self.Intrgl_Fwr_Op(m=m)
 
-                fwr_d = np.zeros(3*self.survey.nRx)
-
-            else:
-
-                print("""Flag must be either 'z' | 'xyz', please revised""")
-                return
-
-            # Add counter to dsiplay progress. Good for large problems
-            count = -1
-            for ii in range(ndata):
-
-                tx, ty, tz = get_T_mat(Xn, Yn, Zn, rxLoc[ii, :])
-
-                if self.rtype == 'z':
-                    fwr_d[ii] = tz.dot(rho)
-
-                elif self.rtype == 'xyz':
-                    fwr_d[ii] = tx.dot(rho)
-                    fwr_d[ii+ndata] = ty.dot(rho)
-                    fwr_d[ii+2*ndata] = tz.dot(rho)
-
-            # Display progress
-                count = progress(ii, count, ndata)
-
-            print("Done 100% ...forward operator completed!!\n")
-
-            return fwr_d
+            return fields
 
         else:
-            return self.G.dot(rho)
-
-    def fields(self, m):
-        self.model = m
+            vec = np.dot(self.F, (self.rhoMap*(m)).astype(np.float32))
 
-        fields = self.fwr_op()
-
-        return fields
-
-    def Jvec(self, m, v, f=None):
-        dmudm = self.rhoMap.deriv(m)
-        return self.G.dot(dmudm*v)
-
-    def Jtvec(self, m, v, f=None):
-        dmudm = self.rhoMap.deriv(m)
-        return dmudm.T * (self.G.T.dot(v))
+            return vec.astype(np.float64)
```

Comment:
@fourndo: can you clarify why we need the `astype` here? it seems like we are switching back and forth between float64 and 32 (line 39) do we need to be doing that?